pkg/quota/controller/clusterquotareconciliation/reconcilation_controller.go

```diff
@@ -0,0 +1,356 @@
+		if err == nil {
+			c.queue.Forget(uncastKey)
+			return false
+		}
+		utilruntime.HandleError(err)
+
+		items := make([]interface{}, 0, len(retryItems))
+		for _, item := range retryItems {
+			items = append(items, item)
+		}
+		c.queue.AddWithDataRateLimited(uncastKey, items...)
+		return false
+	}
+
+	for {
+		if quit := workFunc(); quit {
+			glog.Infof("resource quota controller worker shutting down")
+			return
+		}
+	}
+}
+
+// syncResourceQuotaFromKey syncs a quota key
+func (c *ClusterQuotaReconcilationController) syncQuotaForNamespaces(originalQuota *quotaapi.ClusterResourceQuota, workItems []workItem) (error, []workItem /* to retry */) {
+	obj, err := kapi.Scheme.Copy(originalQuota)
+	if err != nil {
+		return err, workItems
+	}
+	quota := obj.(*quotaapi.ClusterResourceQuota)
+
+	// get the list of namespaces that match this cluster quota
+	matchingNamespaceNamesList, quotaSelector := c.clusterQuotaMapper.GetNamespacesFor(quota.Name)
+	if !kapi.Semantic.DeepEqual(quotaSelector, quota.Spec.Selector) {
+		return fmt.Errorf("mapping not up to date, have=%v need=%v", quotaSelector, quota.Spec.Selector), workItems
+	}
+	matchingNamespaceNames := sets.NewString(matchingNamespaceNamesList...)
+
+	reconcilationErrors := []error{}
+	retryItems := []workItem{}
+	for _, item := range workItems {
+		namespaceName := item.namespaceName
+		namespaceTotals, namespaceLoaded := quota.Status.Namespaces.Get(namespaceName)
+		if !matchingNamespaceNames.Has(namespaceName) {
+			if namespaceLoaded {
+				// remove this item from all totals
+				quota.Status.Total.Used = utilquota.Subtract(quota.Status.Total.Used, namespaceTotals.Used)
+				quota.Status.Namespaces.Remove(namespaceName)
+			}
+			continue
+		}
+
+		// if there's no work for us to do, do nothing
+		if !item.forceRecalculation && namespaceLoaded && kapi.Semantic.DeepEqual(namespaceTotals.Hard, quota.Spec.Quota.Hard) {
+			continue
+		}
+
+		actualUsage, err := quotaUsageCalculationFunc(namespaceName, quota.Spec.Quota.Scopes, quota.Spec.Quota.Hard, c.registry)
+		if err != nil {
+			// tally up errors, but calculate everything you can
+			reconcilationErrors = append(reconcilationErrors, err)
+			retryItems = append(retryItems, item)
+			continue
+		}
+		recalculatedStatus := kapi.ResourceQuotaStatus{
+			Used: actualUsage,
+			Hard: quota.Spec.Quota.Hard,
+		}
+
+		// subtract old usage, add new usage
+		quota.Status.Total.Used = utilquota.Subtract(quota.Status.Total.Used, namespaceTotals.Used)
+		quota.Status.Total.Used = utilquota.Add(quota.Status.Total.Used, recalculatedStatus.Used)
+		quota.Status.Namespaces.Insert(namespaceName, recalculatedStatus)
+	}
+
+	quota.Status.Total.Hard = quota.Spec.Quota.Hard
+
+	if kapi.Semantic.DeepEqual(quota, originalQuota) {
+		return kutilerrors.NewAggregate(reconcilationErrors), retryItems
+	}
+
+	// TODO separate out status updating
+	if _, err := c.clusterQuotaClient.ClusterResourceQuotas().Update(quota); err != nil {
```

Comment:
> every time a clusterquota object is updated (including from a reconcile), that triggers another reconcile?

It triggers the cheap check (It has to), which says, "is this namespace already calculated, yes? skip".  So it triggers the cheap(ish) check, but not the expensive one.